plebnet/agent/qtable.py

```diff
@@ -0,0 +1,158 @@
+            self.environment[provider_of["name"]] = environment_arr
+
+    @staticmethod
+    def calculate_measure(provider_offer):
+        return 1 / (100 * float(provider_offer["price"])) * float(provider_offer["connection"]) * float(
+            provider_offer["memory"])
+
+    def init_providers_offers(self, providers):
+        for i, id in enumerate(providers):
+            options = cloudomate_controller.options(providers[id])
+            for i, option in enumerate(options):
+                element = {
+                    "provider_name": providers[id].get_metadata()[0],
+                    "name": option.name,
+                    "connection": option.connection,
+                    "price": option.price,
+                    "memory": option.memory
+                }
+                self.providers_offers.append(element)
+
+    def update_values(self, status=False):
+        curr_provider = self.self_state["option"]
+        self.update_environment(curr_provider, status)
+
+        for provider_offer in self.providers_offers:
+            for provider_of in self.providers_offers:
+                learning_compound = self.environment[provider_offer["name"]][provider_of["name"]] \
+                                    + self.discount * self.max_action_value(provider_offer) \
+                                    - self.qtable[provider_offer["name"]][provider_of["name"]]
+
+                self.qtable[provider_offer["name"]][provider_of["name"]] = \
+                    self.qtable[provider_offer["name"]][provider_of["name"]] \
+                    + self.learning_rate * learning_compound
+
+    def update_environment(self, provider, status):
+
+        for i, actions in enumerate(self.environment):
+            if not status:
+                self.environment[actions][provider] = self.environment[actions][provider] - self.environment_lr
+            else:
+                self.environment[actions][provider] += self.environment_lr
+
+    def max_action_value(self, provider):
+        max_value = -100000
+        for i, provider_offer in enumerate(self.qtable):
+            if max_value < self.qtable[provider_offer][provider["name"]]:
+                max_value = self.qtable[provider_offer][provider["name"]]
+        return max_value
+
+    def read_dictionary(self, providers=None):
+
+        config_dir = user_config_dir()
+        filename = os.path.join(config_dir, 'QTable.json')
+
+        if not os.path.exists(filename):
+            self.init_qtable_and_environment(providers)
+            self.write_dictionary()
+        else:
+            with open(filename) as json_file:
+                data = json.load(json_file)
+                self.environment = data['environment']
+                self.qtable = data['qtable']
+                self.providers_offers = data['providers_offers']
+                self.self_state = data['self_state']
+
+    def choose_best_option(self, providers):
```

Comment:
Seems very "hard coded"

Reply:
I moved values to the single constant to have less "hard coded code"